### PR TITLE
Move decoders to decoders module

### PR DESCRIPTION
**Pull request by klecki**
Add alias test for image decoder - split decoder is broken.
Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

#### Why we need this PR?
Refactoring to introduce decoders module.

#### What happened in this PR?
 - What solution was applied:
     Introduce alias operators for audio and image* decoders. Move the main ones to `decoders` module
 - Affected modules and functionalities:
     Decoders operators
 - Key points relevant for the review:
     Look out for typos?
 - Validation and testing:
     Sanity alias tests 
 - Documentation (including examples):
     Yes.

**JIRA TASK**: *[DALI-1882]*

### CODE REVIEW

File: dali/operators/decoder/image_decoder.cc [hunk 1]
```diff
@@ -185,8 +185,8 @@ image format, it will decode the entire image and crop the selected ROI.
 
 .. note::
   ROI decoding is currently not compatible with hardware-based decoding. Using
-  :meth:`nvidia.dali.fn.image_decoder` automatically disables hardware accelerated
-  decoding. To use the hardware decoder, use the :meth:`nvidia.dali.fn.image_decoder` and
+  :meth:`nvidia.dali.fn.decoders.image` automatically disables hardware accelerated
```
Comment (jantonguirao) [inline dali/operators/decoder/image_decoder.cc:188]: That's a typo (the doc before). It should be `image_decoder_crop`/`decoders.image_crop`

File: dali/operators/decoder/image_decoder.cc [hunk 2]
```diff
@@ -185,8 +185,8 @@ image format, it will decode the entire image and crop the selected ROI.
...
+  decoding. To use the hardware decoder, use the :meth:`nvidia.dali.fn.decoders.image` and
```
Comment (jantonguirao) [inline dali/operators/decoder/image_decoder.cc:189]: That one is correct

File: dali/operators/decoder/image_decoder.cc [hunk 3]
```diff
@@ -309,4 +309,58 @@ Integer coordinates are interpreted as absolute coordinates, while float coordin
 interpreted as absolute or relative coordinates, depending on the value of
 ``normalized_shape``.)code");
 
+
+// Deprecated aliases
+
+DALI_SCHEMA(ImageDecoder)
+    .DocStr("Legacy alias for :meth:`decoders.image_decoder`.")
```
Comment (jantonguirao) [inline dali/operators/decoder/image_decoder.cc:316]:
```suggestion
    .DocStr("Legacy alias for :meth:`decoders.image`.")
```

File: dali/operators/decoder/image_decoder.cc [hunk 4]
```diff
@@ -309,4 +309,58 @@ Integer coordinates are interpreted as absolute coordinates, while float coordin
...
+    .Deprecate(
+        "decoders__Image",
+        R"code(In DALI 1.0 all decoders were moved into a dedicated :mod:`~nvidia.dali.fn.decoders`
+submodule and renamed to follow a common pattern. This is a placeholder operator with identical
+functionality to allow for backward compatibility.)code");  // Deprecated in 1.0
+
+// Fused
+
+DALI_SCHEMA(ImageDecoderCrop)
+    .DocStr("Legacy alias for :meth:`decoders.image_decoder_crop`.")
```
Comment (jantonguirao) [inline dali/operators/decoder/image_decoder.cc:330]:
```suggestion
    .DocStr("Legacy alias for :meth:`decoders.image_crop`.")
```

File: dali/operators/decoder/image_decoder.cc [hunk 5]
```diff
@@ -309,4 +309,58 @@ Integer coordinates are interpreted as absolute coordinates, while float coordin
...
+    .AddParent("decoders__ImageCrop")
+    .MakeDocPartiallyHidden()
+    .Deprecate(
+        "decoders__ImageCrop",
+        R"code(In DALI 1.0 all decoders were moved into a dedicated :mod:`~nvidia.dali.fn.decoders`
+submodule and renamed to follow a common pattern. This is a placeholder operator with identical
+functionality to allow for backward compatibility.)code");  // Deprecated in 1.0
+
+DALI_SCHEMA(ImageDecoderRandomCrop)
+    .DocStr("Legacy alias for :meth:`decoders.image_decoder_random`.")
```
Comment (jantonguirao) [inline dali/operators/decoder/image_decoder.cc:342]:
```suggestion
    .DocStr("Legacy alias for :meth:`decoders.image_random_crop`.")
```

File: dali/operators/decoder/image_decoder.cc [hunk 6]
```diff
@@ -309,4 +309,58 @@ Integer coordinates are interpreted as absolute coordinates, while float coordin
...
+    .MakeDocPartiallyHidden()
+    .Deprecate(
+        "decoders__ImageRandomCrop",
+        R"code(In DALI 1.0 all decoders were moved into a dedicated :mod:`~nvidia.dali.fn.decoders`
+submodule and renamed to follow a common pattern. This is a placeholder operator with identical
+functionality to allow for backward compatibility.)code");  // Deprecated in 1.0
+
+
+DALI_SCHEMA(ImageDecoderSlice)
+    .DocStr("Legacy alias for :meth:`decoders.image_decoder_slice`.")
```
Comment (jantonguirao) [inline dali/operators/decoder/image_decoder.cc:355]:
```suggestion
    .DocStr("Legacy alias for :meth:`decoders.image_slice`.")
```

File: dali/test/python/test_operator_decoder.py [hunk 7]
```diff
@@ -114,3 +116,55 @@ def check(img_type, size, device, threads):
     for threads in {1, 2, 3, 4}:
         for size in {1, 10}:
             yield check, img_type, size, device, threads
+
+
+
+batch_size_alias_test=16
+
+@pipeline_def(batch_size=batch_size_alias_test, device_id=0, num_threads=4)
+def decoder_pipe(decoder_op, file_root, device, use_fast_idct, split_stages):
```
Comment (jantonguirao) [inline dali/test/python/test_operator_decoder.py:125]: :+1: For testing split_stages

Comment (klecki) [reply dali/test/python/test_operator_decoder.py:125]: If only it would work.

File: dali/operators/decoder/image_decoder.cc (hunk 1 above)
Comment (klecki) [reply dali/operators/decoder/image_decoder.cc:188]: done

File: dali/operators/decoder/image_decoder.cc (hunk 3 above)
Comment (klecki) [reply dali/operators/decoder/image_decoder.cc:316]: done for all

File: dali/operators/decoder/image_decoder.cc (hunk 4 above)
Comment (klecki) [reply dali/operators/decoder/image_decoder.cc:330]: done

File: dali/operators/decoder/image_decoder.cc (hunk 5 above)
Comment (klecki) [reply dali/operators/decoder/image_decoder.cc:342]: done

File: dali/operators/decoder/image_decoder.cc (hunk 6 above)
Comment (klecki) [reply dali/operators/decoder/image_decoder.cc:355]: done

**Comment by klecki** [comment]
!build

File: dali/test/python/test_operator_decoder.py [hunk 8]
```diff
@@ -114,3 +118,62 @@ def check(img_type, size, device, threads):
...
+        new_pipe = decoder_slice_pipe(new_op, file_root, device, use_fast_idct)
+        legacy_pipe = decoder_slice_pipe(old_op, file_root, device, use_fast_idct)
+        compare_pipelines(new_pipe, legacy_pipe, batch_size_alias_test, 10)
+
+def test_image_decoder_slice_alias():
+    data_path = os.path.join(test_data_root, good_path, "jpeg")
+    new_op, old_op = fn.decoders.image_slice, fn.image_decoder_slice
+    for device in ["cpu", "mixed"]:
+        for use_fast_idct in [True, False]:
+                    yield check_image_decoder_slice_alias, new_op, old_op, data_path, device, use_fast_idct
```
Comment (szalpal) [inline dali/test/python/test_operator_decoder.py:179]: Peculiar indent

File: dali/test/python/test_operator_decoder.py [hunk 9]
```diff
@@ -114,3 +118,62 @@ def check(img_type, size, device, threads):
...
+    end = types.Constant(np.array([0.5, 0.5]))
+    decoded = decoder_op(encoded, start, end, device=device,
+                         output_type=types.RGB, use_fast_idct=use_fast_idct)
+    return decoded
+
+
+def check_image_decoder_slice_alias(new_op, old_op, file_root, device, use_fast_idct):
+        new_pipe = decoder_slice_pipe(new_op, file_root, device, use_fast_idct)
+        legacy_pipe = decoder_slice_pipe(old_op, file_root, device, use_fast_idct)
+        compare_pipelines(new_pipe, legacy_pipe, batch_size_alias_test, 10)
```
Comment (szalpal) [inline dali/test/python/test_operator_decoder.py:172]: Peculiar indent

File: dali/test/python/test_operator_decoder.py [hunk 10]
```diff
@@ -114,3 +118,62 @@ def check(img_type, size, device, threads):
...
+
+def test_image_decoder_alias():
+    data_path = os.path.join(test_data_root, good_path, "jpeg")
+    for new_op, old_op in [(fn.decoders.image, fn.image_decoder),
+                           (fn.decoders.image_crop, fn.image_decoder_crop),
+                           (fn.decoders.image_random_crop, fn.image_decoder_random_crop)]:
+        for device in ["cpu", "mixed"]:
+            for use_fast_idct in [True, False]:
+                for split_stages in [False]:
+                        yield check_image_decoder_alias, new_op, old_op, data_path, device, use_fast_idct, split_stages
```
Comment (szalpal) [inline dali/test/python/test_operator_decoder.py:148]: Peculiar indent

File: dali/test/python/test_operator_decoder.py [hunk 11]
```diff
@@ -114,3 +118,62 @@ def check(img_type, size, device, threads):
...
+    encoded, labels = fn.readers.file(file_root=file_root)
+    decoded = decoder_op(encoded, device=device, output_type=types.RGB, use_fast_idct=use_fast_idct,
+                         split_stages=split_stages, seed=42)
+    return decoded
+
+
+def check_image_decoder_alias(new_op, old_op, file_root, device, use_fast_idct, split_stages):
+        new_pipe = decoder_pipe(new_op, file_root, device, use_fast_idct, split_stages)
+        legacy_pipe = decoder_pipe(old_op, file_root, device, use_fast_idct, split_stages)
+        compare_pipelines(new_pipe, legacy_pipe, batch_size_alias_test, 10)
```
Comment (szalpal) [inline dali/test/python/test_operator_decoder.py:137]: Peculiar indent

**Review by szalpal — APPROVED** [review]
Some indentation issues, please fix them.

Also a comment from my side, if we deprecate some operators, shouldn't we keep them out of the table?
image
Or (at least) notify the user better, that he really shouldn't use them... But that's not a blocker, probably out of scope of this PR

File: dali/test/python/test_operator_decoder.py [hunk 12]
```diff
@@ -114,3 +118,62 @@ def check(img_type, size, device, threads):
...
+    for new_op, old_op in [(fn.decoders.image, fn.image_decoder),
+                           (fn.decoders.image_crop, fn.image_decoder_crop),
+                           (fn.decoders.image_random_crop, fn.image_decoder_random_crop)]:
+        for device in ["cpu", "mixed"]:
+            for use_fast_idct in [True, False]:
+                for split_stages in [False]:
+                        yield check_image_decoder_alias, new_op, old_op, data_path, device, use_fast_idct, split_stages
+
+def test_image_decoder_split_alias():
+    raise SkipTest("Sanity tests for aliases of split decoder are temporarily disabled due to accuracy issues.")
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/test/python/test_operator_decoder.py:151]: This is the way. `@nottest` have totally different meaning than skipping a test.

File: dali/test/python/test_operator_decoder.py (hunk 7 above)
Comment (jantonguirao) [reply dali/test/python/test_operator_decoder.py:127]: :)

**Comment by klecki** [comment]
As for the table - I'd say it's fine. It says that this is `legacy alias` and points you to the correct operator. There is also the whole explanation by each of the operators and you get a warning with the same message.

**Review by jantonguirao — APPROVED** [review]

File: dali/test/python/test_operator_decoder.py (hunk 11 above)
Comment (klecki) [reply dali/test/python/test_operator_decoder.py:137]: done

File: dali/test/python/test_operator_decoder.py (hunk 10 above)
Comment (klecki) [reply dali/test/python/test_operator_decoder.py:148]: done

File: dali/test/python/test_operator_decoder.py (hunk 9 above)
Comment (klecki) [reply dali/test/python/test_operator_decoder.py:172]: done

File: dali/test/python/test_operator_decoder.py (hunk 8 above)
Comment (klecki) [reply dali/test/python/test_operator_decoder.py:179]: done

**Review by jantonguirao — APPROVED** [review]

**Pull request merged**